gsk4/src/render_node.rs

```diff
@@ -0,0 +1,173 @@
+            let mut bounds = graphene::Rect::uninitialized();
+            ffi::gsk_render_node_get_bounds(self.to_glib_none().0, bounds.to_glib_none_mut().0);
+            bounds
+        }
+    }
+
+    pub fn get_node_type(&self) -> RenderNodeType {
+        unsafe { from_glib(ffi::gsk_render_node_get_node_type(self.to_glib_none().0)) }
+    }
+
+    pub fn serialize(&self) -> glib::Bytes {
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file<P: AsRef<Path>>(&self, filename: P) -> Result<(), glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+            let _ = ffi::gsk_render_node_write_to_file(
+                self.to_glib_none().0,
+                filename.as_ref().to_glib_none().0,
+                &mut error,
+            );
+            if error.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(error))
+            }
+        }
+    }
+}
+
+pub trait IsARenderNode:
+    StaticType
+    + FromGlibPtrFull<*mut ffi::GskRenderNode>
+    + std::convert::AsRef<crate::RenderNode>
+    + 'static
+{
+    const NODE_TYPE: RenderNodeType;
+    fn upcast(self) -> RenderNode;
+    fn upcast_ref(&self) -> &RenderNode;
+}
+
+pub const NONE_RENDER_NODE: Option<&RenderNode> = None;
+
+impl fmt::Display for RenderNode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("RenderNode")
+    }
+}
+
+#[macro_export]
+macro_rules! define_render_node {
+    ($rust_type:ident, $ffi_type:path, $get_type:path, $node_type:path) => {
+        // Can't use get_type here as this is not a boxed type but another fundamental type
+        glib::wrapper! {
+            #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+            pub struct $rust_type(Shared<$ffi_type>);
+
+            match fn {
+                ref => |ptr| ffi::gsk_render_node_ref(ptr as *mut ffi::GskRenderNode) as *mut $ffi_type,
+                unref => |ptr| ffi::gsk_render_node_unref(ptr as *mut ffi::GskRenderNode),
+            }
+        }
+
+        impl ::glib::StaticType for $rust_type {
+            fn static_type() -> ::glib::Type {
+                unsafe { from_glib($get_type()) }
+            }
+        }
+
+        impl std::convert::AsRef<crate::RenderNode> for $rust_type {
+            fn as_ref(&self) -> &crate::RenderNode {
+                &*self
+            }
+        }
+
+        impl std::ops::Deref for $rust_type {
+            type Target = crate::RenderNode;
+
+            fn deref(&self) -> &Self::Target {
+                unsafe {
+                    &*(self as *const $rust_type as *const crate::RenderNode)
+                }
+            }
+        }
+
+        impl crate::render_node::IsARenderNode for $rust_type {
+            const NODE_TYPE: RenderNodeType = $node_type;
+
+            fn upcast(self) -> crate::RenderNode {
+                unsafe { ::std::mem::transmute(self) }
```

Comment:
If you prefer you can also transmute in `downcast`, or do the `from_glib_full(to_glib_full)` also here. Let's just keep it consistent in both places.

Reply:
I don't know how to use transmute on downcast, since it returns a T.  I just changed it back to `from_glib_full(to_glib_full)` so it can be just like `ParamSpec`.

Reply:
Ok, all the same really as this is not a performance-critical path :)